source/SH3/system/glvertarray.cpp

```diff
@@ -0,0 +1,72 @@
+/** @file
+ *
+ *  Imeplementation of @ref glvertarray.hpp
+ *
+ *  @copyright 2017 Palm Studios
+ *
+ *  @date 8-6-2017
+ *
+ *  @author Jesse Buhagiar
+ */
+#include "SH3/system/glvertarray.hpp"
+#include "SH3/system/log.hpp"
+
+#define GL_VAO_UNBIND 0
+#define BUFFER_OFFSET(i) ((char *)NULL + (i))
+
+using namespace sh3_gl;
+
+void sh3_glvertarray::Create()
+{
+    GLenum err;
+
+    glGenVertexArrays(1, &id);
+
+}
+
+void sh3_glvertarray::Release()
+{
+    glDeleteVertexArrays(1, &id);
+}
+
+void sh3_glvertarray::Bind()
+{
+    GLenum err;
+
+    glBindVertexArray(id);
+
+    if(indexed)
+        buffers.at(Attribute::SH3_INDEX_SLOT).Bind(); // Bind the IBO
+}
+
+void sh3_glvertarray::Unbind()
+{
+    glBindVertexArray(GL_VAO_UNBIND);
+}
+
+void sh3_glvertarray::Draw()
+{
+    Bind();
+    if(indexed)
+        glDrawElements(GL_TRIANGLES, buffers.at(Attribute::SH3_INDEX_SLOT).GetNumberElements(), GL_UNSIGNED_INT, 0);
+    else
+        glDrawArrays(GL_TRIANGLES, 0, buffers.at(Attribute::SH3_VERTEX_SLOT).GetNumberElements());
+    Unbind();
+}
+
+void sh3_glvertarray::SetDataLocation(Attribute attrib, DataType type, GLint size, GLsizei stride, GLint offset)
+{
+    Bind();
+    glEnableVertexAttribArray(attrib); // Set attrib index as currently being used
+    buffers.at(attrib).Bind(); // Bind VBO we are going to send to
+    glVertexAttribPointer(attrib, size, type, GL_FALSE, stride, BUFFER_OFFSET(offset));
+    Unbind();
+}
+
+void sh3_glvertarray::AddVBO(sh3_glbuffer& buff)
+{
+    if(buffers.size() + 1 == 4) // If we have 4 VBOs, we know for sure we have an indexed VAO (Slot 1 VERTS, Slot 2 NORMALS, Slot 3 UV, Slot 4 INDEX)
```

Comment:
Wouldn't that imply that the next size would be 5, and the index buffer has already been added, hence crashing OpenGL?